cmake/CheckHeaders.cmake

```diff
@@ -0,0 +1,113 @@
+# a poor's alternative of iwyu
+#
+# Header files should be self-contained. In general, the source files should
+# not have to adhere to special conditions to include them. For instance,
+# they don't need to include other header files for using a header file, or
+# to define certain macro(s) for using it. But the macros are allowed to be
+# used to its behavior though.
+#
+# include-what-you-use is able to check if all symbols used by a certain header
+# are defined by the headers *directly* included by that header by parsing the
+# header source file using libclang. But we don't always have iwyu installed.
+# Instead of relying on a tool, a new CMake function named
+# "seastar_check_self_contained()" is created to perform a minimal check on the
+# specified set of header files by compiling each of them.
+#
+# Please note, "seastar_check_self_contained()" does not validate the
+# "self-contained" criteria defined above, it just check if the header file
+# *contains* the declarations of the symbols so that any source file including
+# it can be compiled as well. The check performed by the CMake function allows
+# the indirect inclusion of the used symbols. For instance, if "juno.h"
+# references a symbol named "Jupiter", which is declared in "jupiter.h". So,
+# strictly speaking, "juno.h" should include "jupiter.h" for accessing this
+# symbol. But "juno.h" happens to include "solar.h", which in turn includes
+# "jupiter.h". So "seastar_check_self_contained()" accepts "juno.h", while a
+# tool like iwyu would complain at seeing it.
```

Comment:
I'm sorry, but I don't see the problem in this situation. The header is still self-contained. I don't know if "iwyu" complains about this case, but even if it does I don't understand why we should be sad that our tool doesn't.
In fact I think that being overzealos here could add a lot of false-positives - like cases where a header deliberately includes some high-level header which includes a lot of other stuff, and eliminate the need to include a dozen other tiny headers. What's wrong with that? It's even normal practice in a lot of cases (e.g., boost, which in some cases has tiny headers but also bigger headers which include the tiny headers).

Reply:
> I'm sorry, but I don't see the problem in this situation. The header is still self-contained. 

self-contained in a fragile way. then we rely on the implementation of the included headers.

> I don't know if "iwyu" complains about this case, 

it does. that's why i put this explanations here.

> but even if it does I don't understand why we should be sad that our tool doesn't.

explained above.

> In fact I think that being overzealos here could add a lot of false-positives - like cases where a header deliberately includes some high-level header which includes a lot of other stuff, and eliminate the need to include a dozen other tiny headers. What's wrong with that? It's even normal practice in a lot of cases (e.g., boost, which in some cases has tiny headers but also bigger headers which include the tiny headers).

yeah. both boost and we use umbrella headers for this purpose. again, i wanted to explain why we have this and how it is different from a "overzealos" criteria. if you don't think this document helps or wrong or just misleading, i will drop it.

Reply:
Yes, that's what I understood from your comments (and now your reply made it even clearer). It just sounded to me like your comment "apologizes" for not doing the same as that "iwyu" tool, and I don't think we need to apologize. If anything, it sounds to me like this iwyu thing is doing things wrong ;-)

But I agree that you can leave the explanation of how it's different - in case someone is familiar with what iwyu does.

Reply:
> Yes, that's what I understood from your comments (and now your reply made it even clearer). It just sounded to me like your comment "apologizes" for not doing the same as that "iwyu" tool, and I don't think we need to apologize. If anything, it sounds to me like this iwyu thing is doing things wrong ;-)
> 

yeah. a mild "apologizes" is the message i wanted to send. as i wanted to get things right as much as possible. but i thought it's not absolutely right.

> But I agree that you can leave the explanation of how it's different - in case someone is familiar with what iwyu does.

i will 

- remove the paragraph starting with "include-what-you-use". and
- keep "Instead of relying on a tool" one
- keep "Please note"
- remove "If you have include-what-you-use installed"

does it make more sense? hopefully this can help reduce the level of apology to a degree that one can hardly feel. =)